nucypher/blockchain/eth/agents.py

```diff
@@ -1601,6 +1605,103 @@ def execute(self,
         return receipt
 
 
+class InstanceAgent(EthereumContractAgent):
+    """Base class for agents that interact with contracts instantiated by address"""
+
+    def __init__(self, address: str, abi: List[Dict] = None, provider_uri: str = None):
+        blockchain = BlockchainInterfaceFactory.get_or_create_interface(provider_uri=provider_uri)
+        if abi:
+            contract = blockchain.client.w3.eth.contract(abi=abi, address=address)
+        else:  # FIXME: This only works if the blockchain is a deployer interface (i.e. has the solidity compiler)
+            contract = blockchain.get_contract_instance(contract_name=self.contract_name, address=address)
+        super().__init__(contract=contract, provider_uri=provider_uri)
+
+
+class AragonAgent(InstanceAgent):
+    """Base agent for Aragon contracts that uses ABIs from artifact.json files"""
+
+    def __init__(self, address: str, provider_uri: str = None):
+        self.artifact = Artifact(name=self.contract_name)
+        super().__init__(abi=self.artifact.abi, address=address, provider_uri=provider_uri)
+
+
+class ForwarderAgent(AragonAgent):
+    """Agent for Aragon apps that implement the Forwarder interface"""
+
+    contract_name = FORWARDER_INTERFACE_NAME
+
+    def _forward(self, callscript: bytes) -> ContractFunction:  # TODO: Generalize this approach
+        return self.contract.functions.forward(callscript)
+
+    @contract_api(TRANSACTION)
+    def forward(self, callscript: bytes, sender_address: ChecksumAddress) -> TxReceipt:
+        receipt = self.blockchain.send_transaction(contract_function=self._forward(callscript),
+                                                   sender_address=sender_address)
+        return TxReceipt(receipt)
+
+
+class VotingAgent(ForwarderAgent):
+    """Agent to interact with Aragon's Voting apps"""
+
+    contract_name = VOTING_CONTRACT_NAME
+
+    def _new_vote(self,
+                  callscript: bytes,
+                  metadata: str,
+                  cast_vote: bool = None,
+                  execute_if_decided: bool = None) -> ContractFunction:
+        if cast_vote is None and execute_if_decided is None:
+            return self.contract.functions.newVote(callscript, metadata)
+        elif cast_vote is not None and execute_if_decided is not None:
+            return self.contract.functions.newVote(callscript, metadata, cast_vote, execute_if_decided)
+        else:
+            raise ValueError
```

Comment:
not obvious why have two flags with the same check, could you elaborate?

Reply:
This is a way to handle method overloading between the Solidity and the Python layer. In the former, method overloading is simpler as there's no optional parameters: the `newVote` method can be called with either 2 or 4 parameters. Python, however, is less strict, hence I introduced these simple checks to validate the correct number of parameters, since additionally I don't want to be opinionated on default values. I added an error message to clarify.